crates/chain/src/indexed_tx_graph.rs

```diff
@@ -218,9 +218,16 @@ where
             indexer.merge(self.index.index_tx(tx));
         }
 
+        // Create a temporary graph to detect conflicting transactions.
+        let mut temp_graph = self.graph.clone();
+        let _ = temp_graph.batch_insert_unconfirmed(txs.clone());
+
         let graph = self.graph.batch_insert_unconfirmed(
             txs.into_iter()
-                .filter(|(tx, _)| self.index.is_tx_relevant(tx))
+                .filter(|(tx, _)| {
+                    (self.index.is_tx_relevant(tx))
+                        || temp_graph.direct_conflicts(tx).next().is_some()
+                })
                 .map(|(tx, seen_at)| (tx.clone(), seen_at)),
         );
```

Comment:
```rust
        let graph = self.graph.batch_insert_unconfirmed(
            txs.into_iter()
                .filter(|(tx, _)| {
                    (self.index.is_tx_relevant(tx))
                        || self.graph.direct_conflicts(tx).next().is_some()
                })
                .map(|(tx, seen_at)| (tx.clone(), seen_at))
                .collect::<Vec<_>>(),
        );
```

Reply:
If we collect before calling `batch_insert_unconfirmed`, we don't have to hold the immutable reference across it.

Reply:
The `temp_graph` is used here to accurately detect conflicts. Without it, the filtering in this statement happens before any transactions are added to self.graph, causing `self.graph.direct_conflicts()` to always return false.

Reply:
@LagginTimes You'll need to include this in the comment since it's not immediately obvious.

> causing self.graph.direct_conflicts() to always return false.

I don't think this is true in every case. If the indexed tx (the tx which has an output which we own) already exists in the graph, then this will work properly.

I think what you want to say is: This allows us to detect replacement txs which reverts sending to a tracked spk in the situation where the original tx and the replacement tx is added in the same call to `batch_insert_relevant_unconfirmed`.

However, the logic is not fully sound here. Imagine if `batch_insert_relevant_unconfirmed` is called with two conflicting txs which both don't spend or spend to a tracked spk. We expect that this method should filter both these txs out.

Reply:
```rust
        // For any transaction that is relevant by means of tracked spks, we keep track of their
        // spends. It is possible that such a transaction can be replaced with one which no longer
        // includes a tracked spk. We want to include this replacement transaction to prove that
        // the old transaction is evicted.
        let mut relevant_spends = HashSet::<OutPoint>::new();
        let mut indexer_changeset = I::ChangeSet::default();
        for (tx, _) in &txs {
            let new_changeset = self.index.index_tx(tx);
            if !new_changeset.is_empty() {
                indexer_changeset.merge(new_changeset);
                relevant_spends.extend(tx.input.iter().map(|txin| txin.previous_output));
            }
        }

        let tx_graph_changeset = self.graph.batch_insert_unconfirmed(
            txs.into_iter()
                .filter(|(tx, _)| {
                    self.index.is_tx_relevant(tx)
                        || tx
                            .input
                            .iter()
                            .any(|txin| relevant_spends.contains(&txin.previous_output))
                        || self.graph.direct_conflicts(tx).next().is_some()
                })
                .map(|(tx, seen_at)| (tx.clone(), seen_at))
                .collect::<Vec<_>>(),
        );
```